server/store/middleware/infohash/blacklist.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2016 The Chihaya Authors. All rights reserved.
+// Use of this source code is governed by the BSD 2-Clause license,
+// which can be found in the LICENSE file.
+
+package infohash
+
+import (
+	"github.com/chihaya/chihaya"
+	"github.com/chihaya/chihaya/server/store"
+	"github.com/chihaya/chihaya/tracker"
+)
+
+func init() {
+	tracker.RegisterAnnounceMiddleware("infohash_blacklist", blacklistAnnounceInfohash)
+	tracker.RegisterScrapeMiddlewareConstructor("infohash_blacklist", blacklistScrapeInfohash)
+}
+
+// ErrBlockedInfohash is returned by a middleware if any of the infohashes
+// contained in an announce or scrape are disallowed.
+var ErrBlockedInfohash = tracker.ClientError("disallowed infohash")
+
+// blacklistAnnounceInfohash provides a middleware that only allows announces
+// for infohashes that are not stored in a StringStore
+func blacklistAnnounceInfohash(next tracker.AnnounceHandler) tracker.AnnounceHandler {
+	return func(cfg *chihaya.TrackerConfig, req *chihaya.AnnounceRequest, resp *chihaya.AnnounceResponse) (err error) {
+		blacklisted, err := store.MustGetStore().HasString(store.PrefixInfohash + string(req.InfoHash))
+
+		if err != nil {
+			return err
+		} else if blacklisted {
+			return ErrBlockedInfohash
+		}
+		return next(cfg, req, resp)
+	}
+}
+
+// blacklistScrapeInfohash provides a middleware constructor that returns a
+// a middleware that modifies scrape requests based on the infohashes scraped.
```

Comment:
This statement is almost a tautology. This is a middleware constructor that returns a middleware that fails scrape requests to blacklisted infohashes.